### DISCUSSION TITLE
How to trim (remove) white border around structure in picture

### DISCUSSION THREAD

**Discussion by ddsneo4j** [rdkit/rdkit#5879] [Development]
Hi,

  I am using the following code:

from rdkit.Chem import Draw
from rdkit.Chem import AllChem

rxn = AllChem.ReactionFromRxnBlock(r)
d2d = Draw.MolDraw2DCairo(800,400)
d2d.DrawReaction(rxn,highlightByReactant=True,highlightColorsReactants=[(1,1,1),] )
png = d2d.GetDrawingText()
open('mypic','wb+').write(png)

The generated picture  

<img width="300" alt="Test_Output _save" src="https://user-images.githubusercontent.com/42145906/207908071-50324aab-5ea9-4d01-8ca1-75a23842c31f.png">

has a white area around the structure which is too big for my application as it consues to much space on the screen when displaying many structures next to each other.  As a consequence, I would like to trim (i.e. remove) the white space around the structure and just have let's say a 5 pixel white border around it.
Is it possible in RDKit? If yes, instruction steps would be highly appreciated.

Thanks and best regards,
Dan

**Comment by DavidACosgrove** [comment]
Hi @ddsneo4j ,
If you create the MolDraw2DCairo with width and height -1, you will get a canvas that is just big enough to draw a molecule with a fixed scale, with a margin round the outside.  You can control the size of the margin with MolDrawOptions.padding and the scale with MolDrawOptions.scalingFactor (by default it is 20 pixels per molecule unit ("Angstrom", maybe)).
```
from rdkit import Chem
from rdkit.Chem.Draw import rdMolDraw2D, rdDepictor

rdDepictor.SetPreferCoordGen(True)

mol = Chem.MolFromSmiles('C1CCCN1C(C)c1ccc(cc1)c1cnc2c(cnn2c1)c1ccc(S(=O)(=O)N)c2ccccc21')
d = rdMolDraw2D.MolDraw2DSVG(-1, -1)
rdMolDraw2D.SetDarkMode(d)
d.drawOptions().padding = 0.0
d.drawOptions().scalingFactor = 30
rdMolDraw2D.PrepareAndDrawMolecule(d, mol)
d.FinishDrawing()
txt = d.GetDrawingText()
with open('test.svg', 'w') as f:
    f.write(txt)
```
Gives this:

<img width="401" alt="image" src="https://user-images.githubusercontent.com/9198870/207928440-8675f19f-26c6-4228-b798-67fff9fc0385.png">

I have used dark mode and been sloppy with the cut-n-paste of the image so you can see that the molecule goes right up to the edge of the canvas (the black background).  Note that I've used an SVG drawer for no good reason.

HTH,
Dave

**Comment by greglandrum** [comment]
I'm closing this because it is a question, not a bug report or enhancement request.
Please ask questions using the "Discussions" tab.

**Comment by ddsneo4j** [comment]
Hi @DavidACosgrove ,

  thanks for your reply. Please note that I am using reactions. I adapted my code as follows:

        rxn = AllChem.ReactionFromRxnBlock(r)
    
        d2d = Draw.MolDraw2DCairo(-1,-1)
        # use highlighting to avoid images with atom ids, highlighting set to white to be invisible
        d2d.DrawReaction(rxn,highlightByReactant=True,highlightColorsReactants=[(1,1,1),] )
        
        d2d.drawOptions().padding = 0.0
        rdMolDraw2D.PrepareAndDrawMolecule(d2d, rxn)
        d2d.FinishDrawing()

        png = d2d.GetDrawingText()
        name = k[len('Short-RInChIKey='):]
        filename_trim = 'C:\\SRDB\\images_trim\\' + name + '.png'
        with open(filename_trim, 'w') as f:
            f.write(png)


However, I get the following error:

Python argument types in
    rdkit.Chem.Draw.rdMolDraw2D.PrepareAndDrawMolecule(MolDraw2DCairo, ChemicalReaction)
did not match C++ signature:
    PrepareAndDrawMolecule(class RDKit::MolDraw2D {lvalue} drawer, class RDKit::ROMol mol, class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > legend='', class boost::python::api::object highlightAtoms=None, class boost::python::api::object highlightBonds=None, class boost::python::api::object highlightAtomColors=None, class boost::python::api::object highlightBondColors=None, class boost::python::api::object highlightAtomRadii=None, int confId=-1, bool kekulize=True)
Python argument types in
    rdkit.Chem.Draw.rdMolDraw2D.PrepareAndDrawMolecule(MolDraw2DCairo, ChemicalReaction)
did not match C++ signature:
    PrepareAndDrawMolecule(class RDKit::MolDraw2D {lvalue} drawer, class RDKit::ROMol mol, class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > legend='', class boost::python::api::object highlightAtoms=None, class boost::python::api::object highlightBonds=None, class boost::python::api::object highlightAtomColors=None, class boost::python::api::object highlightBondColors=None, class boost::python::api::object highlightAtomRadii=None, int confId=-1, bool kekulize=True)

Do I need to use any other method to fix this?

Thanks and best regards,
Dan

**Reply by greglandrum** [reply]
You are getting that error message because you are calling `DrawMolecule()` with a chemical reaction.

**Reply by ddsneo4j** [reply]
@greglandrum Understood, but what shall I use instead to achieve my goal of trimming the white area around the reaction?
When I omit the PrepareAndDrawMolecule line then I get that:

write() argument must be str, not bytes

**Reply by DavidACosgrove** [reply]
You need to open the file in ‘wb’ mode, as you did in your original code. Mode ‘w’ is for ascii files like svgs.

**Reply by ddsneo4j** [reply]
Thanks @DavidACosgrove , opening in 'wb' mode fixes the "write() argument must be str, not bytes" issue, but the pictures still have the white border that I like to get rid of. I guess I need something equivalent to PrepareAndDrawMolecule (omitted here because it cannot handle reactions, see error below) for drawing reactions?

Python argument types in
    rdkit.Chem.Draw.rdMolDraw2D.PrepareAndDrawMolecule(MolDraw2DCairo, ChemicalReaction)
did not match C++ signature:
    PrepareAndDrawMolecule(class RDKit::MolDraw2D {lvalue} drawer, class RDKit::ROMol mol, class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > legend='', class boost::python::api::object highlightAtoms=None, class boost::python::api::object highlightBonds=None, class boost::python::api::object highlightAtomColors=None, class boost::python::api::object highlightBondColors=None, class boost::python::api::object highlightAtomRadii=None, int confId=-1, bool kekulize=True)
Python argument types in
    rdkit.Chem.Draw.rdMolDraw2D.PrepareAndDrawMolecule(MolDraw2DCairo, ChemicalReaction)
did not match C++ signature:
    PrepareAndDrawMolecule(class RDKit::MolDraw2D {lvalue} drawer, class RDKit::ROMol mol, class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > legend='', class boost::python::api::object highlightAtoms=None, class boost::python::api::object highlightBonds=None, class boost::python::api::object highlightAtomColors=None, class boost::python::api::object highlightBondColors=None, class boost::python::api::object highlightAtomRadii=None, int confId=-1, bool kekulize=True)

**Reply by ddsneo4j** [reply]
> You are getting that error message because you are calling `DrawMolecule()` with a chemical reaction.

@greglandrum Is there any equivalent function for reactions? It seems that in contrast to DrawMolecule() the DrawReaction() method does not respect the -1 setting for height and width that I set in MolDraw2DCairo(-1,-1) to get rid of the white canvas around the reaction.

**Reply by ddsneo4j** [reply]
> > You are getting that error message because you are calling `DrawMolecule()` with a chemical reaction.
> 
> @greglandrum Is there any equivalent function for reactions? It seems that in contrast to DrawMolecule() the DrawReaction() method does not respect the -1 setting for height and width that I set in MolDraw2DCairo(-1,-1) to get rid of the white canvas around the reaction.

@greglandrum Any idea?

**Comment by ddsneo4j** [comment]
Hi @greglandrum,

  sorry, can you move this to discussions then? I am not sure how to do it.

Thanks,
Dan

**Reply by greglandrum** [reply]
> sorry, can you move this to discussions then? I am not sure how to do it.

done

**Reply by ddsneo4j** [reply]
Thanks
